pinot-common/src/main/java/com/linkedin/pinot/common/segment/fetcher/SegmentFetcherFactory.java

```diff
@@ -44,45 +44,57 @@ public static SegmentFetcherFactory getInstance() {
+        put("hdfs", GenericSegmentFetcher.class.getName());
+        put("adl", GenericSegmentFetcher.class.getName());
+        put("file", GenericSegmentFetcher.class.getName());
       }});
   public static final String FETCHER_CLASS_KEY_SUFFIX = ".class";
 
   private final Map<String, SegmentFetcher> _segmentFetcherMap = new HashMap<>();
+  private Configuration _pinotFSConfig;
 
   /**
    * Initiate the segment fetcher factory. This method should only be called once.
+   * @param segmentFetcherClassConfig Segment fetcher factory config
+   * @param pinotFSConfig
    *
-   * @param config Segment fetcher factory config
    */
-  public void init(Configuration config) throws ClassNotFoundException, IllegalAccessException, InstantiationException {
+  public void init(Configuration segmentFetcherClassConfig, Configuration pinotFSConfig) throws ClassNotFoundException, IllegalAccessException, InstantiationException {
     @SuppressWarnings("unchecked")
-    List<String> protocols = config.getList(PROTOCOLS_KEY, DEFAULT_PROTOCOLS);
+    List<String> protocols = segmentFetcherClassConfig.getList(PROTOCOLS_KEY, DEFAULT_PROTOCOLS);
     for (String protocol : protocols) {
       String fetcherClass =
-          config.getString(protocol + FETCHER_CLASS_KEY_SUFFIX, DEFAULT_FETCHER_CLASS_MAP.get(protocol));
+          segmentFetcherClassConfig.getString(protocol + FETCHER_CLASS_KEY_SUFFIX, DEFAULT_FETCHER_CLASS_MAP.get(protocol));
       Preconditions.checkNotNull(fetcherClass, "No fetcher class defined for protocol: " + protocol);
       LOGGER.info("Creating a new segment fetcher for protocol: {} with class: {}", protocol, fetcherClass);
       SegmentFetcher segmentFetcher = (SegmentFetcher) Class.forName(fetcherClass).newInstance();
       LOGGER.info("Initializing segment fetcher for protocol: {}", protocol);
-      Configuration segmentFetcherConfig = config.subset(protocol);
+      Configuration segmentFetcherConfig = segmentFetcherClassConfig.subset(protocol);
       logFetcherInitConfig(segmentFetcher, protocol, segmentFetcherConfig);
       segmentFetcher.init(segmentFetcherConfig);
       _segmentFetcherMap.put(protocol, segmentFetcher);
     }
+    _pinotFSConfig = pinotFSConfig;
   }
 
   public boolean containsProtocol(String protocol) {
     return _segmentFetcherMap.containsKey(protocol);
   }
 
-  public SegmentFetcher getSegmentFetcherBasedOnURI(String uri) throws URISyntaxException {
+  public SegmentFetcher getSegmentFetcherBasedOnURI(String uri) throws Exception {
     String protocol = new URI(uri).getScheme();
-    return _segmentFetcherMap.get(protocol);
+    SegmentFetcher segmentFetcher = _segmentFetcherMap.get(protocol);
+    if (segmentFetcher instanceof GenericSegmentFetcher) {
+      LOGGER.info("Initializing pinotFS for segment fetcher with uri {}", uri);
+      if (_pinotFSConfig.getProperty("file") == null) {
+        _pinotFSConfig.addProperty("file", LocalPinotFS.class.getName());
```

Comment:
Modifying properties within this function seems incorrect. If this is a default property that is expected to be present, lets add it during initialization and not here.